File: Binary Search/MinEatingSpeed.java
class MinEatingSpeed {
  public int minEatingSpeed(int[] piles, int h) {
    int left = 1;
    int right = 0;
    for (int pile : piles)
      right = Math.max(right, pile);
    while (left <= right) {
      int mid = left + (right - left) / 2;
      if (check(mid, piles, h)) {
        right = mid - 1;
      } else {
        left = mid + 1;
      }
    }
    return left;
  }

  public boolean check(double eatingSpeed, int[] piles, int h) {
    int totalHoursToEat = 0;
    for (int pile : piles) {
      double hoursToEatPile = Math.ceil(pile / eatingSpeed);
      totalHoursToEat += hoursToEatPile;
    }

    return totalHoursToEat <= h;
  }

  public static void main(String[] args) {
    MinEatingSpeed solver = new MinEatingSpeed();
    int[] piles = {3, 6, 7, 11};
    int h = 8;
    int result = solver.minEatingSpeed(piles, h);
    System.out.println("Minimum eating speed: " + result);
  }
}

File: BinarySearch/min_eating_speed.py
from typing import List


class MinEatingSpeed:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        def test_eating_speed(speed: int) -> bool:
            total_hours = 0
            for pile in piles:
                total_hours += (pile + speed - 1) // speed
                if total_hours > h:
                    return False
            return True

        left = 1
        right = max(piles)
        while left < right:
            mid = (left + right) // 2
            is_fast_enough = test_eating_speed(mid)
            if is_fast_enough:
                right = mid
            else:
                left = mid + 1

        return left

    # Time complexity: O(n log m), where n is the number of piles 
    # and m is the maximum number of bananas in a pile.
    # The binary search runs in O(log m) time, and for each speed tested, 
    # we iterate through the piles in O(n) time to calculate the total hours needed.
    # Space complexity: O(1)